Fixed and tested the email script

boztopuz opened the pull request.

boztopuz commented:
@ftmzhrsckldr to be able to test the script, I can help you with related credentials.

ftmzhrsckldr commented:
> @ftmzhrsckldr to be able to test the script, I can help you with related credentials.

I have previously written a similar script using smtp, so I know that I need to create an application password and deactivate 2FA to solve the following errors:
``` 
<-- 534-5.7.9 Application-specific password required. For more information, go to
<-- 534 5.7.9  https://support.google.com/mail/?p=InvalidSecondFactor 5b1f17b1804b1-4314209688dsm45218055e9.14 - gsmtp
```
or
```
<-- 535-5.7.8 Username and Password not accepted. For more information, go to
<-- 535 5.7.8  https://support.google.com/mail/?p=BadCredentials 5b1f17b1804b1-4313f56a583sm47826945e9.19 - gsmtp
```

However, when I try to do this for 2024, I encounter the following error:
<img width="1440" alt="Screenshot 2024-10-16 at 15 55 38" src="https://github.com/user-attachments/assets/e357c97c-4d3e-4b30-acde-203797080c55">

How did you pass this error?

boztopuz commented:
@ftmzhrsckldr  To be able to get app passwords you should enable 2FA auth in your google account. After that I search app passwords in google account management

https://capture.dropbox.com/IKDYYqzdCGrzzUZj

After that, I clicked the app passwords and I generated a new app password for script

https://capture.dropbox.com/ETkG9a1dtLVDutOW

ftmzhrsckldr commented on docs/workflows/common-workflow-steps/custom-script.md:
```diff
@@ -227,87 +227,133 @@ sudo -E printenv
+if [ "$os" == "" ]; then
+    echo "This script expects darwin or linux."
+    exit 1
+elif [ "$os" == "darwin" ] && ! which brew > /dev/null 2>&1; then
```
OS is checked in 3 different places for both `darwin` and `linux`. 
This usage does not comply with the **DRY** principle.

> DRY (Don’t repeat yourself)
>
> This principle states that each small pieces of knowledge (code) may only occur exactly once in the entire system.

Could you please edit the code like the pseudocode below?

```bash
...
Create the .msmtprc file

if OS=="darwin":
   if !brew:
      "Error msg: brew not found."
      exit 1
   end
   brew install mailutils & msmtp
   write tls_fingerprint >> ~/.msmtprc
elif OS=="linux":
   if !apt:
      "Error msg: apt not found."
      exit 1
   end
    apt-get install -y mailutils msmtp msmtp-mta
    write ca-cert >> ~/.msmtprc
end
...
```

ftmzhrsckldr commented:
I was able to run this script successfully on "M1 Pool". However, on "Intel Pool" I got the following error:

```
Component Version: 1.0.2
.../step_je3wadfv.e2a/8612d097-3190-47be-b550-7fb54f458ec4.script: line 40: sudo: command not found
.../step_je3wadfv.e2a/8612d097-3190-47be-b550-7fb54f458ec4.script: line 41: sudo: command not found
.../step_je3wadfv.e2a/8612d097-3190-47be-b550-7fb54f458ec4.script: line 87: msmtp: command not found
```

boztopuz commented on docs/workflows/common-workflow-steps/custom-script.md:
```diff
@@ -227,87 +227,133 @@ sudo -E printenv
+
+:::danger Custom Email
+
+To use an SMTP server, this script first installs the necessary certificates, then authenticates to the server with the required credentials and sends the prepared email content to the recipient's email address. In order to change the sender's email address, the SMTP server must allow it by providing the necessary permissions. Otherwise, SMTP servers will send the email using the authenticated em ...
```
@ftmzhrsckldr Actually I think yes, because this script can be used by a person who does not have any idea to work with this script. So, I explain the whole necessary cases.

boztopuz commented:
> @boztopuz I was able to run this script successfully on "M1 Pool". However, on "Intel Pool" I got the following error:
> 
> ```
> Component Version: 1.0.2
> .../step_je3wadfv.e2a/8612d097-3190-47be-b550-7fb54f458ec4.script: line 40: sudo: command not found
> .../step_je3wadfv.e2a/8612d097-3190-47be-b550-7fb54f458ec4.script: line 41: sudo: command not found
> .../step_je3wadfv.e2a/8612d097-3190-47be-b550-7fb54f458ec4.script: line 87: msmtp: command not found
> ```
> 
> Btw I added the comment here because this topic does not have a linear article or I cannot find

@ftmzhrsckldr We can consult to @csonuryilmaz for this. I have not enough knowledge for infrastructure.

ftmzhrsckldr commented:
@boztopuz I tried hard but I couldn't complete my review because you applied the [feedback](https://github.com/appcircleio/appcircle-docusaurus/pull/706#discussion_r1814571286) incorrectly.

Because this feedback was given as **[Pseudocode](https://en.wikipedia.org/wiki/Pseudocode)**. 
This code can't be directly applied to Ruby. And after you made this updates you will need to update the rest of the code because I gave this suggestion to reduce the **if checks** but you didn't make any changes to the other ifs. 

Please ping me after you fix this

ftmzhrsckldr commented:
> > @boztopuz I was able to run this script successfully on "M1 Pool". However, on "Intel Pool" I got the following error:
> > ```
> > Component Version: 1.0.2
> > .../step_je3wadfv.e2a/8612d097-3190-47be-b550-7fb54f458ec4.script: line 40: sudo: command not found
> > .../step_je3wadfv.e2a/8612d097-3190-47be-b550-7fb54f458ec4.script: line 41: sudo: command not found
> > .../step_je3wadfv.e2a/8612d097-3190-47be-b550-7fb54f458ec4.script: line 87: msmtp: command not found
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Btw I added the comment here because this topic does not have a linear article or I cannot find
> 
> @ftmzhrsckldr We can consult to @csonuryilmaz for this. I have not lots of knowledge for infrastructure as well.

@csonuryilmaz, I will not approve this code without your feedback because it **does not work on Intel**. 

If we are not going to make it work on **Intel**, we can at least add a warning that it does not work on **Intel**.

boztopuz commented:
@ftmzhrsckldr I've simplified the if blocks a bit more, I've moved a lot of things into one if block, I have reorganized a whole script. I also tested it with linux_pool in our [dev environment](https://dev-my.appcircle.io/build/detail/ae8b7e3d-ad94-4777-b5ea-a60000cdc36d) and it worked fine there too. 

If I'm still misunderstanding something please pingle, I don't know much about bash and this kind of infra stuff.

ftmzhrsckldr commented on docs/workflows/common-workflow-steps/custom-script.md:
```diff
@@ -227,87 +227,145 @@ sudo -E printenv
-	echo "tls_trust_file /etc/ssl/certs/ca-certificates.crt" >> ~/.msmtprc
-elif [ "$os" == "darwin" ]
-then
-	{ echo -n "tls_fingerprint " &
-	  msmtp --serverinfo --tls --tls-certcheck=off --host=smtp.gmail.com --port=587 \
-	  | egrep -o "([0-9A-Za-z]{2}:){31}[0-9A-Za-z]{2}" ;
-	} >> ~/.msmtprc
+tls on
+EOF
```
Please move this code block after line 260. That is, after the OS check.

ftmzhrsckldr commented on docs/workflows/common-workflow-steps/custom-script.md:
```diff
@@ -227,87 +227,145 @@ sudo -E printenv
+EOF
+
+# Append specific settings for each OS
+if [ "$os" == "darwin" ]; then
+    { echo -n "tls_fingerprint " && msmtp --serverinfo --tls --tls-certcheck=off --host=$HOST_ --port=$PORT_ | egrep -o "([0-9A-Za-z]{2}:){31}[0-9A-Za-z]{2}"; } >> ~/.msmtprc
 fi
```
Please have only one `"$os" == "darwin"` check. So move this change INTO the if conditon at 262

ftmzhrsckldr commented on docs/workflows/common-workflow-steps/custom-script.md:
```diff
@@ -227,87 +227,145 @@ sudo -E printenv
+        echo "Error msg: brew not found."
+        exit 1
+    fi
+    brew install mailutils msmtp
```
The code in the example I gave is pseudocode, that is, code that is not real to explain the algorithm. You shouldn't have taken it directly. Can you please change it back to its old installation ?
```suggestion
    brew install mailutils
    brew install msmtp
```

ftmzhrsckldr commented on docs/workflows/common-workflow-steps/custom-script.md:
```diff
@@ -227,87 +227,145 @@ sudo -E printenv
+    fi
+    brew install mailutils msmtp
+    echo "set sendmail=/usr/local/bin/msmtp" | sudo tee -a /etc/mail.rc
+    echo "tls_fingerprint" >> ~/.msmtprc
```
The code in the example I gave is pseudocode, that is, code that is not real to explain the algorithm. You shouldn't have taken it directly. Can you please change it back to its old code ?

```suggestion
    { echo -n "tls_fingerprint " && msmtp --serverinfo --tls --tls-certcheck=off --host=$HOST_ --port=$PORT_ | egrep -o "([0-9A-Za-z]{2}:){31}[0-9A-Za-z]{2}"; } >> ~/.msmtprc
```

ftmzhrsckldr commented on docs/workflows/common-workflow-steps/custom-script.md:
```diff
@@ -227,87 +227,139 @@ sudo -E printenv
-host smtp.gmail.com
-port 587
+# Create the .msmtprc file with appropriate permissions
+cat <<EOF > ~/.msmtprc
```
In 270 and 279, text is added to the `msmtprc` file, but since ">" is used in the command here, it recreates the content of the `msmtprc` file if it exists.

Therefore, this code should be before the if condition on line 261

ftmzhrsckldr commented:
After this commit [5bcd72e](https://github.com/appcircleio/appcircle-docusaurus/pull/706/commits/5bcd72e1fdcaa7d489fa783214103995e2d230d3) worked on intel and linux

ftmzhrsckldr commented:
@csonuryilmaz I tested this script on Intel pool and M1 pool. There were no errors in my tests.

csonuryilmaz commented on docs/workflows/common-workflow-steps/custom-script.md:
```diff
@@ -227,87 +227,153 @@ sudo -E printenv
+echo "From: $CS_EMAIL_FROM
+To: $CS_EMAIL_TO
+Subject: $CS_EMAIL_SUBJECT
+$CS_EMAIL_BODY" | msmtp --debug --from=$EMAIL_ -t $CS_EMAIL_TO
```
⚠️ `$EMAIL_` is an unassigned variable here.

csonuryilmaz commented on docs/workflows/common-workflow-steps/custom-script.md:
```diff
@@ -227,87 +227,153 @@ sudo -E printenv
+
+# Set email details
+CS_EMAIL_SUBJECT="Test Email Subject"
+CS_EMAIL_TO="reciever-address@mail.com,reciever-2-address@mail.com"
```
We should change the concept from `reciever` to `recipient` since that's the common usage in the email domain. [ref](https://www.google.com/search?q=reciever+email+address)

You can see the list of changes below.

![Screenshot 2024-11-10 at 08 04 32](https://github.com/user-attachments/assets/69e061f7-45fc-497e-bd11-c5dd093a3024)

csonuryilmaz commented on docs/workflows/common-workflow-steps/custom-script.md:
```diff
@@ -227,87 +227,153 @@ sudo -E printenv
+
+:::tip Reciever Email Address
+
+If you want to send an e-mail to multiple e-mail addresses instead of a single e-mail address. In the `CS_EMAIL_TO` parameter, it will be enough to write all the addresses to send an e-mail separated by commas. For example; `CS_EMAIL_TO=example@email.com,example2@email.com`
```
At the beginning of the section, the word was written as `email`, but here we see a different style for the same concept. `e-mail`

Both are correct. ([ref](https://eslbuzz.com/e-mail-or-email/))

On the other hand, **we should have a single way of writing in the section for the same concepts.** Choose one and stick with that.

csonuryilmaz commented on docs/workflows/common-workflow-steps/custom-script.md:
```diff
@@ -227,87 +227,153 @@ sudo -E printenv
-- $PASSWORD_
+- **$CS_EMAIL**: SMTP Server email address.
+- **$CS_USERNAME**: Sender email address.
+- **$CS_PASSWORD**: Sender email address password.
```
For some variables, `$` prefix is not used, but for some of them there is a `$` prefix in writing.

:one: We should have a single style for the same concept. Choose one and stick with that.

:two: Check also other info or tip boxes because I'm not marking them one by one. You can find all of them with short review.
